doc/adr/0007-queryable-properties.md

```diff
@@ -0,0 +1,286 @@
+# 7. Queryable properties
+
+Date: 2023-01-30
+
+## Status
+
+New
+
+## Context
+
+User of the trace-viewer application are often interested in additional information from the data provider. For example, the Trace Compass latency statistics views has columns about minimum and maximum duration and user would like to navigate to the time range where this maximum occurred. The front-end that uses the Trace Server Protocol doesn't know about what data it displaying. It only knows about different graph types. For the latency views it's a table tree with columns, but it doesn't know that there is a minimum column or maximum column, nor it knows where in the trace it occurred.
+
+The Trace Server Protocol (TSP) needs to be augmented to provide such type of data, so that the front-end can provide controls for the user to apply actions in the front-end like select and/or navigate to a given time range, open source code or show a new view using a new custom data provider. These properties are meant to be optional and on-demand, for specific select elements in the views. They should be quick queries and the returned data structures should have limited purpose. They are not meant to query all elements in the view. If that is required, new data providers should be created.
+
+This feature is complementary to the tooltip endpoint for data providers which provides some additional, static details on an UI element. 
+
+### Use cases
+
+- Get min/max time range per entry
+- Get the N longest executions per thread
+- Get crash time (other details like CPU, process, thread)
+- Get most utilized frame (time range)
+- Get source location and open source code file
+- Run critical path analysis and create a data provider descriptor
+
+The challange is to handshake these properties or capabilites between front-end (clients) and server over an domain-agnostic protocol.
+
+### Data Structure
+
+Augment the `DataProviderDescriptor` data structure with information about queryable properties.
+
+```java
+    /*
+        PropertyDescriptor 
+        {
+            name:        string
+            description: string
+            id:          string
+            queryParams: {
```

Comment:
How will the "queryParams" and "metadatakeys" be regulated in the TSP? Will each data provider need to add a description of the available queryParams / properties? In other words, can you provide an example of how the TSP will change too?

Reply:
yeah, I added the metadata part in the last proposal for callsites, because I realized that certain information is not available when constructing the `PropertyDescriptor` and it is dependent on the instance of trace event or state etc. I want to avoid to show or enable a menu if certain information can't be retrieved. For example for callsites, only certain time graph states or trace events can provide that info.

I used metadata because we already use that. The TreeModel for time graphs, xy or data tree can add a metadata map  to each row. For example, the resource view has a metadata map entry for key=cpu and a value. The key is common across views and that's how we are able to select the corresponding CPU row when another view sends a selection signal with such known key.

I agree, it's a bit clunky and I wouldn't have put it in the TSP initially, and only when implementing the solution I would revisit that.  

Hopefully with you counter proposal, it will get also easier to transport callsite information and provide source code look-up for the users in various views. (BTW, in the in the properties view we look for key source to find the source code strin. This was just a prototype and needs to be revisited)

Reply:
Unfortunately I am not fully sure that I have the full understanding of all use cases. I'll try my best to come up with a alternative proposal, but it can take some time, and I do not want to block this proposal if other contributors understand it better than me / feel confident about it.

My greatest concern is that the TSP would become less intuitive to understand, and that possible collaborators would prefer other solutions (e.g. SQL based [perfetto-trace-analysis]) instead of taking advantage of the trace-server features just because the API is too "clunky". That is why I am asking for a pull request in [trace-server-protocol] (connected to this ADR) just to get a better understanding of which endpoints it will impact and how the API will look like if we go this way.

I also would like to make sure that the API we propose will take in consideration:
- if we want to use what you propose here to replace [time-graph-filtering] PR (i.e. use "filter" as one of the possible queryparameters)
- if it makes sense to create our own custom "chart / graph / table query logic", or if we can use something similar to [graphQL]
- that the API we propose is contained in what the trace-server is actually providing, i.e. "data representing charts/graph/tables", and that the API does not diverge too much from that by adding too exotic concepts

[perfetto-trace-analysis]: https://perfetto.dev/docs/quickstart/trace-analysis
[trace-server-protocol]: https://github.com/eclipse-cdt-cloud/trace-server-protocol
[time-graph-filtering]: https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/909
[graphQL]: https://github.com/eclipse-cdt-cloud/theia-trace-extension/blob/master/doc/adr/0002-GraphQL.md

Reply:
> Unfortunately I am not fully sure that I have the full understanding of all use cases. I'll try my best to come up with a alternative proposal, but it can take some time, and I do not want to block this proposal if other contributors understand it better than me / feel confident about it.
> 
I tried to come up with something so that we don't have to come-up with new APIs and endpoints for the different use cases. Maybe it's trying to cover too many cases. Maybe we should just look at the go to min and max and disregard the others. We can put a time range it the tooltip and make it click-able. It could have a hard-carded key value or pattern. 

So, I'm all for simplifying the proposal. I'm looking forward to seeing your alternative proposal. 

> My greatest concern is that the TSP would become less intuitive to understand, and that possible collaborators would prefer other solutions (e.g. SQL based [perfetto-trace-analysis](https://perfetto.dev/docs/quickstart/trace-analysis)) instead of taking advantage of the trace-server features just because the API is too "clunky". That is why I am asking for a pull request in [trace-server-protocol](https://github.com/eclipse-cdt-cloud/trace-server-protocol) (connected to this ADR) just to get a better understanding of which endpoints it will impact and how the API will look like if we go this way.
> 

The ADR has already that information. The sequence diagrams and the data structure definition shows what is added and affected.

> I also would like to make sure that the API we propose will take in consideration:
> 
>     * if we want to use what you propose here to replace [time-graph-filtering](https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/909) PR (i.e. use "filter" as one of the possible queryparameters)
> 

This ADR is not hand;ing the same thing than the other ADR and hence it doesn't replace the other ADR.
>     * if it makes sense to create our own custom "chart / graph / table query logic", or if we can use something similar to [graphQL](https://github.com/eclipse-cdt-cloud/theia-trace-extension/blob/master/doc/adr/0002-GraphQL.md) (and note that here I am talking about embedding graphQL in the trace-server, not to create a "parallel" implementation as the ADR [graphQL](https://github.com/eclipse-cdt-cloud/theia-trace-extension/blob/master/doc/adr/0002-GraphQL.md) proposes)
> 

I'm not clear on how to use GraphQL for this. Please provide a proposal for that so that I can understand.
>     * that the API we propose is contained in what the trace-server is actually providing, i.e. "data representing charts/graph/tables", and that the API does not diverge too much from that by adding too exotic concepts

Still it needs to provide a way to customize for getting additional information.